flow360/component/simulation/validation/validation_simulation_params.py

```diff
@@ -264,3 +268,47 @@ def _validate_cht_has_heat_transfer(params):
                     )
 
     return params
+
+
+def _check_complete_boundary_condition_and_unknown_surface(params):
+    ##:: Step 1: Get all boundaries patches from asset cache
+    asset_boundaries: list[Surface] = params.private_attribute_asset_cache.boundaries
+    if asset_boundaries is None:
+        # No entity info found. Skip the validation.
+        return params
+    ##:: Step 2: Make sure that all boundaries appear in the models section.
+    if params.models is None:
+        raise ValueError("To run a case, `models` field must be defined.")
+    param_boundary_registry = EntityRegistry()
+
+    for model in params.models:
+        if not isinstance(model, get_args(SurfaceModelTypes)):
+            # None-BC models
+            continue
+        # Get referenced entities
+        if hasattr(model, "entities"):
+            # pylint: disable=protected-access
+            for entity in model.entities._get_expanded_entities(create_hard_copy=False):
+                param_boundary_registry.register(entity)
+        elif hasattr(model, "entity_pairs"):  # Periodic BC
+            for surface_pair in model.entity_pairs.items:
+                param_boundary_registry.register(surface_pair.pair[0])
+                param_boundary_registry.register(surface_pair.pair[1])
+
+    for boundary in asset_boundaries:
+        if param_boundary_registry.contains(boundary) is False:
```

Comment:
`.contain` seems be linear and this is also linear loop, which makes it quadratic complexity. `.contain` at least can be improved to be O(log(N)).

Overall I think we can do 2 checks to improve performance here:
1. check for duplication, O(N)
2. check for array equality, we can use Counter as items are hashable and that would be O(N)

so for most checks we will have O(N) cost

only if the above checks fail, we run O(N*log(N)) check to indicate which boundaries are not set or set multiple times.

Reply:
I am confused. Can we have a meeting tomorrow?

Reply:
Using set to speed things up. Checking the length of the two sets does not really help since there might be different items in two sets.